### PR TITLE
BUG: Fix memory leak in dtype from dict contructor

### LINKED ISSUE #13853: Memory leak in recarrays reappeared in 1.16.3
The memory leak on `recarray` which was fixed in 1.15.3 (https://github.com/numpy/numpy/issues/12198) reappeared in 1.16.3+ 

### Reproducing code example:

```python
try:
    import resource  # linux/macos
except ImportError:
    import psutil  # windows
import sys

import numpy as np


def peak_memory_usage():
    """Return peak memory usage in MB"""
    if sys.platform.startswith('win'):
        p = psutil.Process()
        return p.memory_info().peak_wset / 1024 / 1024
    mem = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    factor_mb = 1 / 1024
    if sys.platform == 'darwin':
        factor_mb = 1 / (1024 * 1024)
    return mem * factor_mb


while True:
    np.recarray([1, 2, 3], formats=['i4'])
    print(peak_memory_usage())
```

### Error message:

The number (peak memory usage in MB) is increasing in every iteration...

### Numpy/Python version information:

```
1.16.3 3.7.3 (default, Apr  7 2019, 01:06:41)
[GCC 8.2.1 20181127]
```
Python 3.7.3

### PR DESCRIPTION
**Pull request by seberg**
Fixes gh-13853

### CODE REVIEW

**Comment by seberg** [comment]
Thinking about using `Py_SETREF(new, NULL); goto finish;` instead of that annoying duplication...

File: numpy/core/src/multiarray/descriptor.c [hunk 1]
```diff
@@ -1301,6 +1301,12 @@ _convert_from_dict(PyObject *obj, int align)
             goto fail;
         }
     }
+
+    Py_XDECREF(fields);
+    Py_XDECREF(names);
+    Py_XDECREF(descrs);
+    Py_XDECREF(offsets);
+    Py_XDECREF(titles);
```
Comment (seberg) [inline numpy/core/src/multiarray/descriptor.c]: /home/sebastian/forks/numpy/numpy/core/src/multiarray/descriptor.c

**Comment by seberg** [comment]
Found that strange reference count leak we were looking for Matti ;), since it is trivial pushed a fix (no test)

File: numpy/core/src/multiarray/descriptor.c (hunk 1 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/descriptor.c]: `names` is always null here, it's set on line 1235

Comment (eric-wieser) [reply numpy/core/src/multiarray/descriptor.c]: As for `offsets`, `descrs`, and `titles`, I'd be inclined to decref and null `descrs`, `offsets`, and `titles` on 1212, where they are last used

Comment (seberg) [reply numpy/core/src/multiarray/descriptor.c]: Hmm, maybe better. The nicer pattern would be to use the same cleanup code for success and failure path I think, which is why this unnecessary decref got in. But yeah, I can just decref it all there and think about moving things some other day

Comment (eric-wieser) [reply numpy/core/src/multiarray/descriptor.c]: I suppose I could see it either way. I'll let you make the call.

File: numpy/core/src/common/npy_ctypes.h [hunk 2]
```diff
@@ -32,6 +32,7 @@ npy_ctypes_check(PyTypeObject *obj)
     }
 
     ret = PyObject_IsTrue(ret_obj);
+    Py_DECREF(ret_obj);
```
Comment (eric-wieser) [inline numpy/core/src/common/npy_ctypes.h]: Nice find!

**Review by eric-wieser — APPROVED** [review]

File: numpy/core/src/common/npy_ctypes.h (hunk 2 above)
Comment (seberg) [reply numpy/core/src/common/npy_ctypes.h]: Have a whole bunch more, I think refcount testing might be a big soon (except f2py probably), that would be awesome...

**Pull request merged**